## Overview
In this post, I'll give a gentle introduction to the conditional probability with the help of a real life example. Then, I'll try to extend this idea of calculating conditional probabilities to Bayes Theorem. I know there already exist excellent resources on this but this post is no way a replacement to those, it's just a supplement to see how we can use those concepts to some real-life data. 

In probability theory, conditional probability is the probability of occurring of an event where it is given that another event has already occurred. To understand it a little better, we first need to set up the stage by defining a few terms from set theory.

## Events and Sample Space
In the simplest terms, an event is just the result of a random experiment. For example, getting a head when we toss a coin is one event, drawing a ball at random from a bag containing 3 black and 5 red balls is also an event. As we can see, we can easily associate the concept of probability to the events. 

A collection of all possible outcomes of an event is called a sample space, for tossing the coin we can have just two outcomes: head (H) or a tail(T). Similarly, rolling a fair die will always result in some number between 1 to 6, hence the sample space is {1, 2, 3, 4, 5, 6}.

## Union of events 
Consider again the rolling of a fair die where we define two events:
- __Event A:__ Getting a number which is divisible by 2
- __Event B:__ Getting a number which is divisible by 3
The sample space for event A is {2, 4, 6} whereas for event B it is {3, 6}. Now if we define another event C which is getting a number which is divisible by either 2 __OR__ 3 our new sample space would just be the combination of all the unique elements of sample space A and sample space B: {2, 3, 4, 6}. Mathematically, these events can be shown in terms of venn diagrams:
<a href="https://imgur.com/7o5B3u9"><img src="https://i.imgur.com/7o5B3u9.png" title="source: imgur.com" /></a>

In terms of probabilities, we can easily calculate the probabilities for all the events as follows:
- P(A) = $\frac{\mathrm{Number \:of \:cases \:where \:the \:output \:is \:divisible \:by \:2}}{\mathrm{Total \:possible \:outcomes}}$ = 3/6 = 0.5
- P(B) = $\frac{\mathrm{Number \:of \:cases \:where \:the \:output \:is \:divisible \:by \:3}}{\mathrm{Total \:possible \:outcomes}}$ = 2/6 = 0.333
- P(C) = P(A $\cup$ B) = $\frac{\mathrm{Number \:of \:cases \:where \:the \:output \:is \:divisible \:by \:either \:2 \:OR \:3}}{\mathrm{Total \:possible \:outcomes}}$ = 4/6 = 0.667

## Intersection of events
Following the events defined previously, we can also define an event D which is getting a number which is divisible by both 2 __AND__ 3, meaning _the common element in the sample space of both the events_. In terms of venn diagrams, it can be shown as:
<a href="https://imgur.com/dwBjCji"><img src="https://i.imgur.com/dwBjCji.png" title="source: imgur.com" /></a>

Again, in terms of probabilities:
- P(A) = $\frac{\mathrm{Number \:of \:cases \:where \:the \:output \:is \:divisible \:by \:2}}{\mathrm{Total \:possible \:outcomes}}$ = 3/6 = 0.5
- P(B) = $\frac{\mathrm{Number \:of \:cases \:where \:the \:output \:is \:divisible \:by \:3}}{\mathrm{Total \:possible \:outcomes}}$ = 2/6 = 0.333
- P(C) = P(A $\cap$ B) = $\frac{\mathrm{Number \:of \:cases \:where \:the \:output \:is \:divisible \:by \:both \:2 \:AND \:3}}{\mathrm{Total \:possible \:outcomes}}$ = 1/6 = 0.167

## Disjoint events
Consider these two events:
- __Event A:__ Getting a number which is divisible by 2
- __Event B:__ Getting a number which is divisible by 5
The sample space for event A is {2, 4, 6} whereas for event B it is {5}. We can see that these events cannot occur together in the case of rolling a fair die. These events are called disjoint events. The venn diagram for these type of events can be shown as:
<a href="https://imgur.com/KqXin9c"><img src="https://i.imgur.com/KqXin9c.png" title="source: imgur.com" /></a>

## Dependent and Independent events
If the occurrence of one event does not effect the occurrence of another event,then these events are termed as independent events. Few examples of independent events include:
- Getting a head when a coin is tossed AND getting a 5 in rolling a fair die
- Getting rains in the month of August AND snow in December

The probability in the case of independent events can be written as `P(A `$\cap$` B) = P(A) * P(B)`, that is, the probability of occurring both the events is just the product of individual probabilities. Let's try to understand this more concretely, suppose we have a bag containing 3 blue and 5 green balls and we draw two balls at random with replacement (*meaning putting back the first ball in the bag after first trial*). We define the two events as:

- __Event A:__ Getting a blue ball
- __Event B:__ Getting a green ball

We are interested in calculating the probability of getting a blue ball in the first trial AND a green ball in the second. Writing the probabilities for both events:
`P(A) = 3/8`; `P(B) = 5/8`. Since getting a blue ball in a trial is independent of getting the green ball, this is the case of independent events. So we can write:

P(C) = P(A $\cap$ B) = P(A) * P(B) = P(B) * P(A) = 15 / 64

Now for the case of dependent events, we can use the same example with one big difference: __we are not going to replace the drawn ball in the first trial!__ In this case, `P(A) = 3/8` and `P(B) = 5/7`. Notice the denominator of `P(B)`: since the drawn ball is not replaced, now we are sampling the second draw out of a smaller number of sample space. This would also ensure a greater chance of getting the green ball in the second draw. 

This discussion of dependent events naturally extends to the idea of conditional probability: We try to calculate the probability of an event A given another event B has already happened. It is denoted by `P(A|B)`. To get a feel for it, let's see some examples based on it
- Probability of drawing a diamond from a deck of well-shuffled cards given the drawn card is red
- Probability of rain on a given day of the month given it is July
We can easily infer from above two examples that both the events in the examples are dependent of each other. 

Conditional probability can be defined as follows:

**Probability of event A given event B has already occurred = P(A|B) = $\frac{P(A\cap B)}{P(B)}$**

we can easily see the above equation reduces to `P(A)` for independent events by writing `P(A`$\cap$`B) = P(A)*P(B)`. 

Now in my attempt to make this post a little less boring, let's make our hands dirty and use python to understand the concept of conditional probability.

In [1]:
# Import libraries
import numpy as np
import pandas as pd

## About the Dataset
The dataset contains the monthly rainfall data from years 1901 to 2018 for the Indian state of Kerala. Kerala is one of the few states which are usually badly hit by monsoons every year. You can read more about it in this [excellent kernel](https://www.kaggle.com/biphili/india-rainfall-kerala-flood).

In [2]:
# Read the data
df = pd.read_csv("/kaggle/input/kerela-flood/kerala.csv")
df.head(20)

SUBDIVISION  YEAR   JAN   FEB   MAR    APR    MAY     JUN     JUL    AUG  \
0       KERALA  1901  28.7  44.7  51.6  160.0  174.7   824.6   743.0  357.5   
1       KERALA  1902   6.7   2.6  57.3   83.9  134.5   390.9  1205.0  315.8   
2       KERALA  1903   3.2  18.6   3.1   83.6  249.7   558.6  1022.5  420.2   
3       KERALA  1904  23.7   3.0  32.2   71.5  235.7  1098.2   725.5  351.8   
4       KERALA  1905   1.2  22.3   9.4  105.9  263.3   850.2   520.5  293.6   
5       KERALA  1906  26.7   7.4   9.9   59.4  160.8   414.9   954.2  442.8   
6       KERALA  1907  18.8   4.8  55.7  170.8  101.4   770.9   760.4  981.5   
7       KERALA  1908   8.0  20.8  38.2  102.9  142.6   592.6   902.2  352.9   
8       KERALA  1909  54.1  11.8  61.3   93.8  473.2   704.7   782.3  258.0   
9       KERALA  1910   2.7  25.7  23.3  124.5  148.8   680.0   484.1  473.8   
10      KERALA  1911   3.0   4.3  18.2   51.0  180.6   990.0   705.3  178.6   
11      KERALA  1912   1.9  15.0  11.2  122.7  217.3   948.2   833.6  534.4   
12      KERALA  1913   3.1   5.2  20.7   75.7  198.8   541.7   763.2  247.2   
13      KERALA  1914   0.7   6.8  18.1   32.7  164.2   565.3   857.7  402.2   
14      KERALA  1915  16.9  23.5  42.7  106.0  154.5   696.1   775.6  298.8   
15      KERALA  1916   0.0   7.8  22.0   82.4  199.0   920.2   513.9  396.9   
16      KERALA  1917   2.9  47.6  79.4   38.1  122.9   703.7   342.7  335.1   
17      KERALA  1918  42.9   5.0  32.8   51.3  683.0   464.3   167.5  376.0   
18      KERALA  1919  43.0   6.1  33.9   65.9  247.0   636.8   648.0  484.2   
19      KERALA  1920  35.2   5.5  24.1  172.0   87.7   964.3   940.8  235.0   

      SEP    OCT    NOV    DEC   ANNUAL RAINFALL FLOODS  
0   197.7  266.9  350.8   48.4            3248.6    YES  
1   491.6  358.4  158.3  121.5            3326.6    YES  
2   341.8  354.1  157.0   59.0            3271.2    YES  
3   222.7  328.1   33.9    3.3            3129.7    YES  
4   217.2  383.5   74.4    0.2            2741.6     NO  
5   131.2  251.7  163.1   86.0            2708.0     NO  
6   225.0  309.7  219.1   52.8            3671.1    YES  
7   175.9  253.3   47.9   11.0            2648.3     NO  
8   195.4  212.1  171.1   32.3            3050.2    YES  
9   248.6  356.6  280.4    0.1            2848.6     NO  
10   60.2  302.3  145.7   87.6            2726.7     NO  
11  136.8  469.5  138.7   22.0            3451.3    YES  
12  176.9  422.5  109.9   45.8            2610.8     NO  
13  241.0  374.4  100.9  135.2            2899.1     NO  
14  396.6  196.6  302.5   14.9            3024.5    YES  
15  339.3  320.7  134.3    8.9            2945.3    YES  
16  470.3  264.1  256.4   41.6            2704.8     NO  
17   96.4  233.2  295.4   54.1            2501.9     NO  
18  255.9  249.2  280.1   53.0            3003.3    YES  
19  178.0  350.1  302.3    8.2            3303.1    YES

In [3]:
# Changing the target column to numeric values
df["FLOODS"] = df["FLOODS"].map({"YES": 1, "NO": 0})

df

SUBDIVISION  YEAR   JAN   FEB   MAR    APR    MAY     JUN     JUL     AUG  \
0        KERALA  1901  28.7  44.7  51.6  160.0  174.7   824.6   743.0   357.5   
1        KERALA  1902   6.7   2.6  57.3   83.9  134.5   390.9  1205.0   315.8   
2        KERALA  1903   3.2  18.6   3.1   83.6  249.7   558.6  1022.5   420.2   
3        KERALA  1904  23.7   3.0  32.2   71.5  235.7  1098.2   725.5   351.8   
4        KERALA  1905   1.2  22.3   9.4  105.9  263.3   850.2   520.5   293.6   
..          ...   ...   ...   ...   ...    ...    ...     ...     ...     ...   
113      KERALA  2014   4.6  10.3  17.9   95.7  251.0   454.4   677.8   733.9   
114      KERALA  2015   3.1   5.8  50.1  214.1  201.8   563.6   406.0   252.2   
115      KERALA  2016   2.4   3.8  35.9  143.0  186.4   522.2   412.3   325.5   
116     KERALA   2017   1.9   6.8   8.9   43.6  173.5   498.5   319.6   531.8   
117      KERALA  2018  29.1  52.1  48.6  116.4  183.8   625.4  1048.5  1398.9   

       SEP    OCT    NOV    DEC   ANNUAL RAINFALL  FLOODS  
0    197.7  266.9  350.8   48.4            3248.6       1  
1    491.6  358.4  158.3  121.5            3326.6       1  
2    341.8  354.1  157.0   59.0            3271.2       1  
3    222.7  328.1   33.9    3.3            3129.7       1  
4    217.2  383.5   74.4    0.2            2741.6       0  
..     ...    ...    ...    ...               ...     ...  
113  298.8  355.5   99.5   47.2            3046.4       1  
114  292.9  308.1  223.6   79.4            2600.6       0  
115  173.2  225.9  125.4   23.6            2176.6       0  
116  209.5  192.4   92.5   38.1            2117.1       0  
117  423.6  356.1  125.4   65.1            4473.0       1  

[118 rows x 16 columns]

We will be needing only columns `JUN`, `JUL`, `YEAR` and `FLOODS` since we are interested in calculating the probability of flooding in that year given it rained more than a certain threshold (500 mm) in these months. We will create a couple more columns based on these columns.

In [4]:
# Creating binary data for the months of June and July using the rainfall threshold
df["JUN_GT_500"] = (df["JUN"] > 500).astype("int")
df["JUL_GT_500"] = (df["JUL"] > 500).astype("int")
df_small = df.loc[:, ["YEAR", "JUN_GT_500", "JUL_GT_500", "FLOODS"]]
df_small["COUNT"] = 1
df_small.head()

YEAR  JUN_GT_500  JUL_GT_500  FLOODS  COUNT
0  1901           1           1       1      1
1  1902           0           1       1      1
2  1903           1           1       1      1
3  1904           1           1       1      1
4  1905           1           1       0      1

In [5]:
df_small.shape

(118, 5)

In [6]:
# Creating the tabular data based on the counts
pd.crosstab(df_small["FLOODS"], df_small["JUN_GT_500"])


JUN_GT_500   0   1
FLOODS            
0           19  39
1            6  54

Defining some variables:
- `P(F)`: Probability of flooding
- `P(J)`: Probability of having more than 500 mm rain in June
- `P(F`$\cap$`J)`: Probability of flooding __and__ having more than 500 mm rain in June
- `P(F|J)`: Probability of flooding given it rained more than 500 mm in June

Based on the above table we can easily find these probabilities.

In [7]:
P_F = (6 + 54) / (6 + 54 + 19 + 39)
P_J = (39 + 54) / (6 + 54 + 19 + 39)
P_F_intersect_J = 54 / (6 + 54 + 19 + 39)
print(f"P(F): {P_F}") 
print(f"P(J): {P_J}")
print(f"P(F AND J): {P_F_intersect_J}")

P(F): 0.5084745762711864
P(J): 0.788135593220339
P(F AND J): 0.4576271186440678


Using the formula - `P(A|B) = P(A`$\cap$`B) / P(B)` we can easily calculate the conditional probability:

In [8]:
# Now calculate probailitity of flood given it rained more than 500 mm in June (P(A|B))
P_F_J = P_F_intersect_J / P_J
print(f"P(F|J): {P_F_J}")

P(F|J): 0.5806451612903226


Now, we can also ask ourselves this: <span style="color:red">*given that it flooded in Kerala in a given year what is the probability that it rained more than 500 mm in the month of June or July?*</span> This is where **Bayes Theorem** comes into action. Some other examples of Bayes Theorem are like:
- The probability of a woman having breast cancer given she tested positive in the test
- Probability that a given email is actually a spam given it contains certain flagged words.

Bayes Theorem can be easily derived using the relationship between conditional probability and intersection of events. Given two events, we already know:
$$
P(A \cap B) = P(A|B).P(B) = P(B|A).P(A) \\
\mathrm{so},\: P(B|A) = \frac{P(A|B).P(B)}{P(A)}
$$

<span style="color:blue">In Bayesian inference, `P(B)` is called **Prior Probability**. In our case, `P(J)` is the prior probability which tells the probability of rain more than 500 mm in June (or July) without knowing whether it flooded or not that year. We can see prior probability is the probability of the event we are interested in before any new information.</span>

Okay, enough chatter, let's try to code this in python. Actually we have already done most of the work, it's just a matter of plugging in the numbers into the above equation.

In [9]:
# Probability of rain more than 500 mm in June given it flooded that year (P(B|A))
P_J_F = (P_F_J * P_J) / P_F
print(f"P(J|F): {P_J_F}")

P(J|F): 0.9000000000000001


In [10]:
# We can similarly do it for july
pd.crosstab(df_small["FLOODS"], df_small["JUL_GT_500"])

JUL_GT_500   0   1
FLOODS            
0           19  39
1            3  57

Defining the similar parameters for July:
- `P(F)`: Probability of flooding
- `P(J)`: Probability of having more than 500 mm rain in July
- `P(F`$\cap$`J)`: Probability of flooding __and__ having more than 500 mm rain in July
- `P(F|J)`: Probability of flooding given it rained more than 500 mm in July

In [11]:
P_F = (3 + 57) / (3 + 57 + 19 + 39)
P_J = (39 + 57) / (3 + 57 + 19 + 39)
P_F_intersect_J = 57 / (3 + 57 + 19 + 39)
print(f"P(F): {P_F}") 
print(f"P(J): {P_J}")
print(f"P(F AND J): {P_F_intersect_J}")

P(F): 0.5084745762711864
P(J): 0.8135593220338984
P(F AND J): 0.4830508474576271


In [12]:
# Now calculate probailitity of flood given it rained more than 500 mm in July
P_F_J = P_F_intersect_J / P_J
print(f"P(F|J): {P_F_J}")

P(F|J): 0.59375


In [13]:
# Probability of rain more than 500 mm in July given it flooded that year (P(B|A))
P_J_F = (P_F_J * P_J) / P_F
print(f"P(J|F): {P_J_F}")

P(J|F): 0.9500000000000002


## Important Takeaways
1. Based on the probability outputs above we can easily infer that it flooded almost 59% of the time in the year when it rained more than 500 mm in July whereas for June it's only 58%. This means only rainfall in the months of June and July are not completely responsible for the flooding in Kerala. This actually makes sense since in both 2018 and 2020, the flooding happened in August. May be including August in the analysis provide more insight to this.
2. Using Bayes theorem we found that whenever it flooded in Kerala, both June and July have a very high probability (90% and 95% respectively) of rain for more than 500 mm. This also makes sense June and July are the peak months of rainfall because of monsoon.

References and Resources:
- [https://www.statisticshowto.com/bayes-theorem-problems/](https://www.statisticshowto.com/bayes-theorem-problems/)
- [https://www.analyticsvidhya.com/blog/2017/03/conditional-probability-bayes-theorem/](https://www.analyticsvidhya.com/blog/2017/03/conditional-probability-bayes-theorem/)
- [https://towardsdatascience.com/bayes-theorem-the-holy-grail-of-data-science-55d93315defb](https://towardsdatascience.com/bayes-theorem-the-holy-grail-of-data-science-55d93315defb)